aiohttp/client.py

```diff
@@ -784,11 +775,11 @@ async def _ws_connect(
             extstr = ws_ext_gen(compress=compress)
             real_headers[hdrs.SEC_WEBSOCKET_EXTENSIONS] = extstr
 
-        if not isinstance(ssl, SSL_ALLOWED_TYPES):
-            raise TypeError(
-                "ssl should be SSLContext, bool, Fingerprint, "
-                "or None, got {!r} instead.".format(ssl)
-            )
+        # For the customers that mistakenly pass in True, convert it to None to use default ssl context
+        if ssl is True:
+            ssl = None
```

Comment:
@webknjaz what do you think? Currently, nothing in the documentation mentions `True` as an allowed value. I'm wondering if it just makes more sense to replace `None` with `True` (in v4) and drop `None` as an allowed value entirely. In which case we can just keep the existing SSL type checks instead...

Reply:
At least from a user perspective, allowing `False` but not allowing `True` is a little weird. :)

Reply:
If @webknjaz isn't around, maybe @bdraco can give an opinion on this.

Reply:
@bdraco Thanks for the comments.

Currently, the valid types of ssl parameter are SSLContext, Literal[False], Fingerprint or None.

If user sets ssl = False, we disable ssl certificate validation which makes total sense. But if user set ssl = True by mistake, instead of enabling ssl certificate validation or raising errors, we silently disable the validation too which is a little subtle but weird.

As you said, we want to avoid breaking changes as much as possible.

Hence in this PR, we added a check that if user sets ssl=True, we enable certificate validation by treating it as using Default SSL Context.

Does this make sense? (I also added these into the PR description)

Reply:
That makes sense. `True` definitely doesn't do what one would expect. I'm sure there are cases where people have passed in `True` and hoped it would enable verification.

By changing the meaning of `True`, we introduce a breaking change since they now get verification, which may break their production code that may have relied on no verification (most likely mistakenly). It's a good change, but still a breaking one (even if they shouldn't be doing that).

I think that's OK, but we should warn them about the behavior change so they don't have to dig through the docs (and maybe the code) to determine what has changed. I imagine that would be a difficult change to find without some feedback from the code.

Reply:
Are you leaning towards reverting the type checking and only having the change

if ssl=True:
   ssl=None

?

I am open to it too if @Dreamsorcerer & @webknjaz are ok.

Reply:
After re-read your comments, are you saying that the change may break the case that user explicitly sets the ssl to True but relies on the behavior of no ssl verification?

I agree this is technically breaking. But I assume this is a positive breaking change. :)

Reply:
We should also get a test that setting `True` now enables verification

Reply:
If we want to change it to `True`, then we can always add a DeprecationWarning when using `None`, to give people plenty of time to switch any explicit arguments before v4.

Reply:
> `True` has never had a meaning, it's an invalid input. So, I'd prefer to disallow `True`, or replace `None` with `True` in future releases (bearing in mind that `None` is the default, so may not be passed explicitly by many people).

Raising when passing `True` seems fine to me as well. It makes it easy to see what went wrong.

> If we want to change it to `True`, then we can always add a DeprecationWarning when using `None`, to give people plenty of time to switch any explicit arguments before v4.

The downside to adding support for `True` is that anyone already passing it gets no warning that the behavior has changed because it becomes a valid option after an update, so we can't warn them of a behavior change.

Since previously it was an invalid option and a bug for them to pass it, assuming we are only talking about changing it out from under them if they explicitly pass a value (it should be secure by default with no passed value) to work more like one would intuitively expect (True/False) vs. (None/False) might be worth the cost of the few users who will get no warning.

Reply:
> The downside to adding support for `True` is that anyone already passing it gets no warning that the behavior has changed because it becomes a valid option after an update, so we can't warn them of a behavior change.

This part doesn't concern me, as it is obvious their intention was to enable ssl, i.e. the default behaviour. The behaviour change will merely make it do what they were expecting.

I think the only users who need to update anything are those who get a deprecation warning from passing None (maybe because they wrap the client and have their own default arguments matching ours).

Reply:
If you agree, then I'll leave that with you to oversee. I'll be back Wednesday.

Reply:
> This part doesn't concern me, as it is obvious their intention was to enable ssl, i.e. the default behaviour. The behaviour change will merely make it do what they were expecting.
> 
> I think the only users who need to update anything are those who get a deprecation warning from passing None (maybe because they wrap the client and have their own default arguments matching ours).

> If you agree, then I'll leave that with you to oversee. I'll be back Wednesday.

Sounds good. Let's change it to be `True``/`False` and deprecate explicit `None` (passing nothing should default to `True`)

Reply:
To make sure I understand correctly, we want to change it to:
1. Default the value of ssl to True.
2. If user sets it to True explicitly, we will change the behavior from disabling ssl verification to enabling ssl verfication
3. If user explicitly sets it to None, we will keep current behavior which is to enable ssl verification and add a deprecation warning

Is this correct?

Reply:
That looks correct to me 👍

Reply:
Slight tweak, as I forgot about backports.
So, this PR for master:
Change value to True. Remove None as an allowed value.

Then when that is merged, we can do the backports, I think in 3.9 we should allow None and True. Then in 3.10 backport we add a deprecation warning to None. That should cause the minimal amount of problems.

Reply:
Here's my opinion on the order of the changes: https://github.com/aio-libs/aiohttp/pull/7698#discussion_r1411152254.